source/iterator/snapshot.go

```diff
@@ -0,0 +1,261 @@
+	db             *sqlx.DB
+	table          string
+	orderingColumn string
+	keys           []string
+	batchSize      int
+	position       *position.Position
+	columnTypes    map[string]string
+	suffixName     string
+}
+
+func newSnapshotIterator(
+	ctx context.Context,
+	snapshotParams snapshotParams,
+) (*snapshotIterator, error) {
+	var err error
+
+	it := &snapshotIterator{
+		db:             snapshotParams.db,
+		table:          snapshotParams.table,
+		keys:           snapshotParams.keys,
+		orderingColumn: snapshotParams.orderingColumn,
+		batchSize:      snapshotParams.batchSize,
+		position:       snapshotParams.position,
+		columnTypes:    snapshotParams.columnTypes,
+		suffixName:     snapshotParams.suffixName,
+	}
+
+	err = it.loadRows(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("load rows: %w", err)
+	}
+
+	if snapshotParams.position != nil {
+		it.maxValue = snapshotParams.position.SnapshotMaxValue
+	} else {
+		err = it.setMaxValue(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("set max value: %w", err)
+		}
+	}
+
+	return it, nil
+}
+
+// HasNext check ability to get next record.
+func (i *snapshotIterator) HasNext(ctx context.Context) (bool, error) {
+	if i.rows != nil && i.rows.Next() {
+		return true, nil
+	}
+
+	if err := i.loadRows(ctx); err != nil {
+		return false, fmt.Errorf("load rows: %w", err)
+	}
+
+	// check new batch.
+	if i.rows != nil && i.rows.Next() {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// Next get new record.
+func (i *snapshotIterator) Next(ctx context.Context) (sdk.Record, error) {
+	row := make(map[string]any)
+	if err := i.rows.MapScan(row); err != nil {
+		return sdk.Record{}, fmt.Errorf("scan rows: %w", err)
+	}
+
+	transformedRow, err := columntypes.TransformRow(ctx, row, i.columnTypes)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("transform row column types: %w", err)
+	}
+
+	if _, ok := transformedRow[i.orderingColumn]; !ok {
+		return sdk.Record{}, ErrNoOrderingColumn
+	}
+
+	pos := position.Position{
+		IteratorType:             position.TypeSnapshot,
+		SnapshotLastProcessedVal: transformedRow[i.orderingColumn],
+		SnapshotMaxValue:         i.maxValue,
+		SuffixName:               i.suffixName,
+	}
+
+	sdkPos, err := pos.ConvertToSDKPosition()
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("convert position %w", err)
+	}
+
+	keysMap := make(map[string]any)
+	for _, val := range i.keys {
+		if _, ok := transformedRow[val]; !ok {
+			return sdk.Record{}, fmt.Errorf("key %v, %w", val, ErrNoKey)
+		}
+
+		keysMap[val] = transformedRow[val]
+	}
+
+	transformedRowBytes, err := json.Marshal(transformedRow)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("marshal row: %w", err)
+	}
+
+	i.position = &pos
+
+	metadata := sdk.Metadata(map[string]string{metadataTable: i.table})
+	metadata.SetCreatedAt(time.Now())
+
+	return sdk.Util.Source.NewRecordSnapshot(
+			sdkPos,
+			metadata,
+			sdk.StructuredData(keysMap),
+			sdk.RawData(transformedRowBytes)),
```

Comment:
because conduit can return error if we structured data and  value with time.Time type in payload
Something like it
```
source stream was stopped unexpectedly: error reading from source: read stream error: error converting key: proto: invalid type: time.Time
```

Reply:
There it says "error converting key", and not payload though? We use structured data in Postgres as well (and that _should_ work).

Reply:
ok, I will retest it

Reply:
This is quite unexpected...

Reply:
Confirmed it's an issue in Conduit: https://github.com/ConduitIO/conduit/issues/866